lib/meadow/pipeline/dispatcher.ex

```diff
@@ -0,0 +1,123 @@
+defmodule Meadow.Pipeline.Actions.Dispatcher do
+  @moduledoc """
+  Action responsible for dispatching processing to
+  the next action in the pipeline
+
+  Subscribes to: all other actions with status: :ok
+  *
+
+  """
+  alias Meadow.Data.{ActionStates, FileSets}
+  alias Meadow.Utils.Atoms
+  alias Sequins.Pipeline.Action
+
+  alias Meadow.Pipeline.Actions.{
+    CopyFileToPreservation,
+    CreatePyramidTiff,
+    ExtractMimeType,
+    FileSetComplete,
+    GenerateFileSetDigests,
+    ExtractExifMetadata,
+    IngestFileSet,
+    InitializeDispatch
+  }
+
+  use Action
+  require Logger
+
+  @actiondoc "Disatch pipeline actions specific to role + mime_type"
+
+  @initial_actions [
+    IngestFileSet,
+    ExtractMimeType,
+    InitializeDispatch
+  ]
+
+  @access_image_actions [
+    GenerateFileSetDigests,
+    ExtractExifMetadata,
+    CopyFileToPreservation,
+    CreatePyramidTiff,
+    FileSetComplete
+  ]
+
+  @preservation_actions [
+    GenerateFileSetDigests,
+    CopyFileToPreservation,
+    FileSetComplete
+  ]
+
+  @preservation_image_actions [
+    GenerateFileSetDigests,
+    ExtractExifMetadata,
+    CopyFileToPreservation,
+    FileSetComplete
+  ]
+
+  def process(data, attributes) do
+    Logger.info("dispatcher called")
+    file_set = FileSets.get_file_set(data.file_set_id)
+
+    dispatch_next_action(file_set, attributes)
+    :ok
+  end
+
+  def process(data, attributes, _), do: process(data, attributes)
+
+  def initial_actions, do: @initial_actions
+
+  def all_progress_actions do
+    Enum.uniq(
+      @initial_actions ++
+        @preservation_actions ++ @preservation_image_actions ++ @access_image_actions
+    )
+  end
+
+  def dispatcher_actions(%{role: %{id: "A"}, core_metadata: %{mime_type: "image/" <> _}}),
+    do: @access_image_actions
+
+  def dispatcher_actions(%{role: %{id: "X"}, core_metadata: %{mime_type: "image/" <> _}}),
+    do: @access_image_actions
+
+  def dispatcher_actions(%{role: %{id: "P"}, core_metadata: %{mime_type: "image/" <> _}}),
+    do: @preservation_image_actions
+
+  def dispatcher_actions(%{role: %{id: "S"}}),
+    do: @preservation_actions
+
+  def dispatcher_actions(_), do: nil
+
+  def not_my_actions(file_set) do
+    (all_progress_actions() -- @initial_actions) -- dispatcher_actions(file_set)
+  end
+
+  defp dispatch_next_action(file_set, attributes) do
+    last = ActionStates.get_latest_state(file_set.id)
```

Comment:
You can get the name of the last action (and its completion status) straight out of the `attributes` hash:

```
%{process: "CreatePyramidTiff", role: "A", status: "ok"}
```

(These are the same attributes referenced in `sequins.exs` that sequins uses to create subscriptions.)

I think the advantage to using the `process` attribute here is for when we might want to manually run a file set back through one or more actions. `ActionStates.get_latest_state/1` has a better chance of being confused by the existence of other actions, whereas `process` will tell you exactly what just finished. You just have to use 

```
Module.safe_concat(Meadow.Pipeline.Actions, action_name)
```

to get the fully qualified module name of the action (e.g., to turn the string `"CreatePyramidTiff"` into a reference to the module `Meadow.Pipeline.Actions.CreatePyramidTiff`).